Issue #144 solved by injecting cosmosdb sdk object using asp.net core dependency injection framework

ritesh-modi opened the pull request:
# Type of PR

- [ ] Documentation changes
- [X] Code changes
- [ ] Test changes
- [ ] CI-CD changes
- [ ] GitHub Template changes

## Purpose of PR

Injecting cosmosdb client object using asp.net core dependency injection framework

## Validation

- [ ] Unit tests updated and ran successfully
- [ ] Update documentation or issue referenced above

## Issues Closed or Referenced

- Closes #144 (this will automatically close the issue when the PR closes)
- References #144 (this references the issue but does not close with PR)

Commit:
* Program.cs:
* appsettings.json:
* GlobalSuppressions.cs:
* dalMain.cs:
* DalExtension.cs:
* KeyRotationHelper.cs:

* idal.cs:

changed code to use asp.net core dependency frameowrk to
  create and use CosmosClient object

* Constants.cs:

ritesh-modi committed:
Merge pull request #1 from retaildevcrews/main

pulling latest code from upstream

Commit:
Issue #144 solved by injecting cosmosdb sdk object using asp.net core dependency injection framework

bartr commented on src/app/appsettings.json:
```diff
@@ -6,5 +6,10 @@
       "Microsoft": "Error"
     }
+  },
+  "CosmosOptions": {
```
we should add these via system.commandline instead of appsettings for consistency

ritesh-modi replied:
I am now reusing the existing CosmosConfig class instead of appsettings.  Removed entries from appsettings file.

bartr commented on src/app/Program.cs:
```diff
@@ -238,66 +244,5 @@ private static async Task<IWebHost> BuildHost(string kvUrl, AuthenticationType a
             // build the host
             return builder.Build();
         }
-
```
Is there a reason you deleted this? It's still needed for "proactive" key rotation vs. "reactive"

ritesh-modi replied:
I have added it back to the code. My thought was that it was being refactored within its own class. If we have to remove the code again, we can do so later.

bartr commented on src/app/CSE.KeyRotation/KeyRotationHelper.cs:
```diff
@@ -52,8 +58,17 @@ private AsyncRetryPolicy GetCosmosRetryPolicy()
+                            MaxRetryWaitTimeOnRateLimitedRequests = TimeSpan.FromSeconds(configuration.GetValue<int>(Constants.CosmosTimeout)),
+                        };
+
+                        services.AddSingleton<CosmosClient>(new CosmosClient(configuration[Constants.CosmosUrl], cosmosKeySecret.Value));
```
it seems strange to have this code here instead of startup or program.cs - is it possible to move it for consistency?

ritesh-modi replied:
This code was added here to replace the existing cosmos client object within the dependency container with the new one since there is a change in the key.

bartr commented on src/app/DataAccessLayer/DataAccessLayerExtension.cs:
```diff
@@ -19,15 +20,32 @@ public static class DataAccessLayerExtension
+                MaxRetryWaitTimeOnRateLimitedRequests = TimeSpan.FromSeconds(timeout),
+            };
+
+            services.AddSingleton<CosmosClient>(new CosmosClient(cosmosUrl.AbsoluteUri, cosmosKey, options));
```
I would build an extension so that you can call services.AddCosmosClient() or something similar - for ease of use and consistency

ritesh-modi replied:
added a new extension class and file

aflinchb reviewed (commented):
Could you remove the .DS_Store files and also add that to .gitignore so they aren't accidentally added in the future?

Commit:
* appsettings.json: removed appsettings configuration specific to
  cosmosclient object

* KeyRotationHelper.cs: code refactoring

* CosmosClientExtension.cs:
* DataAccessLayerExtension.cs: added new extension file and code for
  cosmos client. refactored code

* Program.cs: incorporated changes based on comments and feedback.
  added back key rotation function

* GlobalSuppressions.cs: added suppressmessage for not disposing
  cosmosclient object. It should not be disposed.
